File: test_messagecontextinfo_fix.js
#!/usr/bin/env node
/**
 * Test to verify that messages with messageContextInfo are not filtered out
 * This is a critical fix for WhatsApp message handling
 */

// Read the baileys_service.js file and extract the extractText function
const fs = require('fs');
const path = require('path');

const serviceFile = path.join(__dirname, 'services', 'whatsapp', 'baileys_service.js');
const content = fs.readFileSync(serviceFile, 'utf8');

// Extract the extractText function using regex
const extractTextMatch = content.match(/function extractText\(msgObj\) \{[\s\S]*?\n\}/);
if (!extractTextMatch) {
  console.error('❌ Could not find extractText function');
  process.exit(1);
}

// Create a safe eval environment
const extractTextCode = extractTextMatch[0];
eval(extractTextCode);

console.log('🧪 Testing messageContextInfo fix...\n');

// Test Case 1: Message with conversation and messageContextInfo should NOT be filtered
const testCase1 = {
  conversation: "נראלך לא נספיק לסיים עד מחר?",
  messageContextInfo: {
    deviceListMetadata: {},
    deviceListMetadataVersion: 2
  }
};

const result1 = extractText(testCase1);
console.log('Test 1: Message with conversation and messageContextInfo');
console.log(`  Input: conversation="${testCase1.conversation}", messageContextInfo=present`);
console.log(`  Expected: "${testCase1.conversation}"`);
console.log(`  Got: "${result1}"`);

if (result1 === testCase1.conversation) {
  console.log('  ✅ PASS - Message correctly extracted\n');
} else {
  console.log('  ❌ FAIL - Message was incorrectly filtered\n');
  process.exit(1);
}

// Test Case 2: Message with ONLY messageContextInfo should still be filtered
const testCase2 = {
  messageContextInfo: {
    deviceListMetadata: {},
    deviceListMetadataVersion: 2
  }
};

const result2 = extractText(testCase2);
console.log('Test 2: Message with ONLY messageContextInfo (no actual content)');
console.log(`  Input: messageContextInfo=present, no text content`);
console.log(`  Expected: null`);
console.log(`  Got: ${result2}`);

if (result2 === null) {
  console.log('  ✅ PASS - Non-message correctly filtered\n');
} else {
  console.log('  ❌ FAIL - Non-message should have been filtered\n');
  process.exit(1);
}

// Test Case 3: Regular message without messageContextInfo should work
const testCase3 = {
  conversation: "Hello, this is a test"
};

const result3 = extractText(testCase3);
console.log('Test 3: Regular message without messageContextInfo');
console.log(`  Input: conversation="${testCase3.conversation}"`);
console.log(`  Expected: "${testCase3.conversation}"`);
console.log(`  Got: "${result3}"`);

if (result3 === testCase3.conversation) {
  console.log('  ✅ PASS - Regular message works correctly\n');
} else {
  console.log('  ❌ FAIL - Regular message failed\n');
  process.exit(1);
}

// Test Case 4: Protocol message should be filtered (even with text)
const testCase4 = {
  protocolMessage: {
    type: 0
  },
  conversation: "This should be filtered"
};

const result4 = extractText(testCase4);
console.log('Test 4: Protocol message should be filtered');
console.log(`  Input: protocolMessage=present, conversation="${testCase4.conversation}"`);
console.log(`  Expected: null`);
console.log(`  Got: ${result4}`);

if (result4 === null) {
  console.log('  ✅ PASS - Protocol message correctly filtered\n');
} else {
  console.log('  ❌ FAIL - Protocol message should have been filtered\n');
  process.exit(1);
}

console.log('🎉 All tests passed!');
console.log('\n✅ FIX VERIFIED: Messages with messageContextInfo are now correctly processed');

File: test_messagecontextinfo_fix.py
#!/usr/bin/env python3
"""
Test to verify that messages with messageContextInfo are not filtered out.
This is a critical fix for WhatsApp message handling.

The fix: messageContextInfo is metadata that can accompany real messages,
so it should NOT cause messages to be filtered out.
"""

import sys
import os

sys.path.insert(0, '/home/runner/work/prosaasil/prosaasil')

def test_messagecontextinfo_not_in_filter():
    """Verify messageContextInfo is not in the filter list"""
    print("🧪 Testing messageContextInfo fix in baileys_service.js\n")
    
    service_file = '/home/runner/work/prosaasil/prosaasil/services/whatsapp/baileys_service.js'
    
    with open(service_file, 'r') as f:
        content = f.read()
    
    # Find the extractText function
    start_idx = content.find('function extractText(msgObj) {')
    if start_idx == -1:
        print("❌ FAIL - Could not find extractText function")
        return False
    
    # Get the first 500 characters of the function (should contain the filter logic)
    function_snippet = content[start_idx:start_idx + 500]
    
    # Check that messageContextInfo is NOT in the filter condition
    filter_start = function_snippet.find('if (msgObj.pollUpdateMessage')
    filter_end = function_snippet.find('return null;  // Ignore silently')
    
    if filter_start == -1 or filter_end == -1:
        print("❌ FAIL - Could not find filter condition")
        return False
    
    filter_condition = function_snippet[filter_start:filter_end]
    
    print("Filter condition found:")
    print(filter_condition)
    print()
    
    if 'messageContextInfo' in filter_condition:
        print("❌ FAIL - messageContextInfo is still in the filter condition")
        print("This will cause messages with messageContextInfo to be incorrectly filtered out")
        return False
    else:
        print("✅ PASS - messageContextInfo is NOT in the filter condition")
        print("Messages with messageContextInfo will now be processed correctly")
        print()
    
    # Check that other filters are still in place
    required_filters = [
        'pollUpdateMessage',
        'protocolMessage',
        'historySyncNotification',
        'reactionMessage',
        'senderKeyDistributionMessage'
    ]
    
    missing_filters = []
    for filter_name in required_filters:
        if filter_name not in filter_condition:
            missing_filters.append(filter_name)
    
    if missing_filters:
        print(f"⚠️  WARNING - Missing expected filters: {missing_filters}")
        print("Make sure these system message types are still being filtered")
        return False
    else:
        print("✅ PASS - All expected system message filters are still in place")
        print()
    
    # Check that there's a comment explaining the fix
    if 'Note: messageContextInfo is metadata' in content[start_idx:start_idx + 800]:
        print("✅ PASS - Comment explaining the fix is present")
    else:
        print("⚠️  INFO - No comment found explaining why messageContextInfo is not filtered")
    
    print()
    return True

def test_conversation_field_is_checked():
    """Verify that the conversation field is properly checked for text content"""
    print("🧪 Testing that conversation field is checked for text extraction\n")
    
    service_file = '/home/runner/work/prosaasil/prosaasil/services/whatsapp/baileys_service.js'
    
    with open(service_file, 'r') as f:
        content = f.read()
    
    # Find the extractText function
    start_idx = content.find('function extractText(msgObj) {')
    if start_idx == -1:
        print("❌ FAIL - Could not find extractText function")
        return False
    
    # Get a larger snippet to include the conversation check
    function_snippet = content[start_idx:start_idx + 1500]
    
    # Check that conversation field is being checked
    if 'if (msgObj.conversation)' in function_snippet or 'msgObj.conversation' in function_snippet:
        print("✅ PASS - Conversation field is checked for text content")
        return True
    else:
        print("❌ FAIL - Conversation field check not found")
        return False

def main():
    print("=" * 70)
    print("WhatsApp Message Filtering Fix - Validation")
    print("=" * 70)
    print()
    
    results = []
    
    # Test 1: messageContextInfo is not in filter
    results.append(test_messagecontextinfo_not_in_filter())
    
    # Test 2: conversation field is still checked
    results.append(test_conversation_field_is_checked())
    
    print("=" * 70)
    if all(results):
        print("🎉 ALL TESTS PASSED")
        print()
        print("✅ FIX VERIFIED:")
        print("   - Messages with messageContextInfo will now be processed")
        print("   - System messages are still correctly filtered")
        print("   - Text extraction logic is intact")
        return 0
    else:
        print("❌ SOME TESTS FAILED")
        return 1

if __name__ == '__main__':
    sys.exit(main())
